File: frontend/src/lib/types/target.ts
import { TaskStatus } from './task-status';
import type { WhoisLookupType } from './whois';

export enum TargetType {
	DOMAIN = 'domain',
	IP = 'ip',
	IP_RANGE = 'ip_range',
	ASN = 'asn',
	URL = 'url'
}

export interface WhoisSummaryData {
	id: string;
	query_value: string;
	lookup_type: WhoisLookupType;
	name: string;
	registrant_name: string;
	registrar_name: string;
	country: string;
	network_cidr: string;
	registration_date: string | null;
	expiration_date: string | null;
	queried_at: string;
}

export interface DnsSummaryData {
	id: string;
	host: string;
	status_code: string;
	cdn: boolean;
	cdn_name: string;
	queried_at: string;
	record_counts: Record<string, number>;
}

export interface BgpSummaryData {
	// ASN targets
	prefix_count: number | null;
	peer_count: number | null;
	announced: boolean | null;

	// IP / IP_RANGE targets
	asn: number | null;
	prefix: string | null;
	holder: string | null;

	queried_at: string;
}

export interface OrganizationSummary {
	id: string;
	name: string;
	slug: string;
}

export interface TagSummary {
	id: string;
	name: string;
	slug: string;
	color: string;
}

export interface TargetBase {
	target_value: string;
	display_name?: string | null;
}

export interface Target extends TargetBase {
	id: string;
	target_type: TargetType;
	project_id: string;
	created_at: string;
	updated_at: string;
	created_by: string;
	whois_status: TaskStatus;
	whois_error: string | null;
	whois_record_id: string | null;
	whois: WhoisSummaryData | null;
	bgp_status: TaskStatus;
	bgp: BgpSummaryData | null;
	dns_status: TaskStatus;
	dns_error: string | null;
	dns_lookup_id: string | null;
	dns: DnsSummaryData | null;
	organizations: OrganizationSummary[];
	tags: TagSummary[];
}

export interface TargetCreate extends TargetBase {
	project_slug: string;
	organization_names?: string[];
	tag_names?: string[];
}

export interface TargetUpdate {
	display_name?: string | null;
	organization_names?: string[] | null;
	tag_names?: string[] | null;
}

export interface TargetValidationRequest {
	target_value: string;
}

export interface TargetValidationResponse {
	valid: boolean;
	target_type: TargetType | null;
	error: string | null;
	target_value: string;
}

export interface TargetBulkCreateRequest {
	project_slug: string;
	targets: string[];
	organization_names?: string[];
	tag_names?: string[];
}

export interface TargetImportItem {
	target_value: string;
	tags?: string[];
	organizations?: string[];
	display_name?: string | null;
}

export interface TargetImportRequest {
	project_slug: string;
	targets: TargetImportItem[];
}

export interface TargetImportResult {
	target_value: string;
	success: boolean;
	target_type: TargetType | null;
	target_id: string | null;
	error: string | null;
}

export interface TargetBulkCreateResponse {
	total: number;
	imported: number;
	failed: number;
	skipped_duplicates: number;
	results: TargetImportResult[];
}

export interface TargetPreviewItem {
	target_value: string;
	target_type?: TargetType | null;
	tags?: string[];
	organizations?: string[];
	display_name?: string | null;
	error?: string;
}

export interface EnrichmentRefreshResponse {
	target_id: string;
	enrichment_type: string;
	status: string;
	message: string;
}

export function getTargetTypeColor(_type: TargetType): string {
	return 'bg-muted/60 text-foreground/70 border-border/60';
}

export function formatTargetType(type: TargetType): string {
	return type.replace('_', ' ').toUpperCase();
}

export const TARGET_ASSET_NOUN: Record<TargetType, string> = {
	[TargetType.DOMAIN]: 'subdomain',
	[TargetType.IP]: 'host',
	[TargetType.IP_RANGE]: 'host',
	[TargetType.ASN]: 'host',
	[TargetType.URL]: 'host'
};

export function targetAssetNoun(type: string, count = 2): string {
	const noun = TARGET_ASSET_NOUN[type as TargetType] ?? 'host';
	return count === 1 ? noun : `${noun}s`;
}

File: shared/models/target.py
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from pydantic import BaseModel
from sqlmodel import Field, Relationship, SQLModel

from shared.enums.target import TargetType
from shared.enums.whois import WhoisStatus
from shared.models.organization import Organization, OrganizationSummary
from shared.models.tag import TagSummary, TargetTag
from shared.models.whois import WhoisRecord, WhoisRecordSummary
from shared.utils.datetime import utc_now

if TYPE_CHECKING:
    from shared.models.tag import Tag


class TargetValidationRequest(BaseModel):
    target_value: str


class TargetValidationResponse(BaseModel):
    valid: bool
    target_type: TargetType | None
    error: str | None
    target_value: str


class TargetOrganization(SQLModel, table=True):
    __tablename__ = "target_organizations"

    target_id: uuid.UUID = Field(foreign_key="targets.id", primary_key=True)
    organization_id: uuid.UUID = Field(foreign_key="organizations.id", primary_key=True)


class TargetBase(SQLModel):
    target_value: str = Field(max_length=500)
    display_name: str | None = Field(default=None, max_length=200)


class Target(TargetBase, table=True):
    __tablename__ = "targets"
    __table_args__ = ({"sqlite_autoincrement": True},)

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True, index=True)
    target_type: TargetType
    project_id: uuid.UUID = Field(foreign_key="projects.id", index=True)
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)
    created_by: uuid.UUID = Field(foreign_key="users.id")

    whois_status: WhoisStatus = Field(default=WhoisStatus.PENDING, index=True)
    whois_error: str | None = Field(default=None, max_length=1000)
    whois_record: WhoisRecord | None = Relationship(
        sa_relationship_kwargs={"lazy": "selectin"},
    )
    whois_record_id: uuid.UUID | None = Field(
        default=None, foreign_key="whois_records.id", index=True
    )

    organizations: list["Organization"] = Relationship(
        link_model=TargetOrganization,
        sa_relationship_kwargs={"lazy": "selectin"},
    )

    tags: list["Tag"] = Relationship(
        back_populates="targets",
        link_model=TargetTag,
        sa_relationship_kwargs={"lazy": "selectin"},
    )


class TargetCreate(TargetBase):
    project_slug: str
    organization_names: list[str] = Field(default_factory=list)
    tag_names: list[str] = Field(default_factory=list)


# bulk targets create/request/response models
class TargetBulkCreate(BaseModel):
    project_slug: str
    targets: list[str] = Field(
        ...,
        min_length=1,
        max_length=1000,
        description="List of target values to import (max 1000)",
    )
    organization_names: list[str] = Field(default_factory=list)
    tag_names: list[str] = Field(default_factory=list)


class TargetImportResult(BaseModel):
    # single target import result
    target_value: str
    success: bool
    target_type: TargetType | None = None
    target_id: uuid.UUID | None = None
    error: str | None = None


class TargetBulkCreateResponse(BaseModel):
    total: int
    imported: int
    failed: int
    skipped_duplicates: int
    results: list[TargetImportResult]


class TargetUpdate(SQLModel):
    display_name: str | None = Field(default=None, max_length=200)
    organization_names: list[str] | None = None
    tag_names: list[str] | None = None


class TargetRead(TargetBase):
    id: uuid.UUID
    target_type: TargetType
    project_id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID
    whois_status: WhoisStatus
    whois_error: str | None
    whois_record_id: uuid.UUID | None
    whois: WhoisRecordSummary | None = None
    organizations: list[OrganizationSummary] = Field(default_factory=list)
    tags: list[TagSummary] = Field(default_factory=list)


# imports for both json and csv support
class TargetImportItem(BaseModel):
    target_value: str
    tags: list[str] = Field(default_factory=list)
    organizations: list[str] = Field(default_factory=list)
    display_name: str | None = None


class TargetImportRequest(BaseModel):
    project_slug: str
    targets: list[TargetImportItem] = Field(
        ...,
        min_length=1,
        max_length=500,
        description="List of targets to import (max 500)",
    )
